### PR TITLE
Correctly check types in type assertions

### LINKED ISSUE #9103: Empty tuples are allowed in type assertions
**TypeScript Version:** 

master

**Code**

```ts
var a: [] = <[]>[];
```

**Expected behavior:**

```
$ node built/local/tsc.js index.ts
index.ts(1,8): error TS1122: A tuple type element list cannot be empty.
index.ts(1,14): error TS1122: A tuple type element list cannot be empty.
```

**Actual behavior:** 

```
$ node built/local/tsc.js index.ts
index.ts(1,8): error TS1122: A tuple type element list cannot be empty.
```

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
Fixes #9103.

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -5027,6 +5027,7 @@ namespace ts {
         function getTypeFromTupleTypeNode(node: TupleTypeNode): Type {
             const links = getNodeLinks(node);
             if (!links.resolvedType) {
+                checkTupleType(node);
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: That's not the right fix. This function is simply for materializing the type, it isn't for the actual checking. The real problem is that `checkAssertion` never calls `checkSourceElement` on `node.type`. You can prove that to yourself by observing that this doesn't generate an error either:

```ts
var x = <{ a: number, a: number}>undefined;  // No duplicate identifier error
```

So, remove the line above and instead put a call to `checkSourceElement(node.type)` in `checkAssertion`.

**Comment by DanielRosenwasser** [comment]
Fixed.

**Comment by mhegazy** [comment]
👍

File: tests/baselines/reference/defaultValueInFunctionTypes.errors.txt [hunk 2]
```diff
@@ -1,8 +1,11 @@
...
 
 
-==== tests/cases/compiler/defaultValueInFunctionTypes.ts (1 errors) ====
+==== tests/cases/compiler/defaultValueInFunctionTypes.ts (2 errors) ====
     var x: (a: number = 1) => number;
             ~~~~~~~~~~~~~
 !!! error TS2371: A parameter initializer is only allowed in a function or constructor implementation.
-    var y = <(a : string = "") => any>(undefined)
\ No newline at end of file
+    var y = <(a : string = "") => any>(undefined)
```
Comment (RyanCavanaugh) [inline tests/baselines/reference/defaultValueInFunctionTypes.errors.txt]: :laughing:

**Comment by RyanCavanaugh** [comment]
:+1:

**Pull request merged**